coffee/rewriter.py

```diff
@@ -131,132 +145,131 @@ def licm(self, merge_and_simplify=False, compact_tmps=False):
+        self.expr_expander.expand(should_expand)
 
         # Update known declarations
-        self.decls.update(ee.expanded_decls)
-        self._expanded = True
+        self.decls.update(self.expr_expander.expanded_decls)
 
-    def distribute(self):
-        """Factorize terms in the expression.
-        E.g. ::
+    def factorize(self, mode='standard'):
+        """Factorize terms in the expression. For example: ::
 
             A[i]*B[j] + A[i]*C[j]
 
         becomes ::
 
-            A[i]*(B[j] + C[j])."""
-
-        def find_prod(node, occs, to_distr):
-            if isinstance(node, Par):
-                find_prod(node.children[0], occs, to_distr)
-            elif isinstance(node, Sum):
-                find_prod(node.children[0], occs, to_distr)
-                find_prod(node.children[1], occs, to_distr)
-            elif isinstance(node, Prod):
-                left, right = (node.children[0], node.children[1])
-                l_str, r_str = (str(left), str(right))
-                if occs[l_str] > 1 and occs[r_str] > 1:
-                    if occs[l_str] > occs[r_str]:
-                        dist = l_str
-                        target = (left, right)
-                        occs[r_str] -= 1
-                    else:
-                        dist = r_str
-                        target = (right, left)
-                        occs[l_str] -= 1
-                elif occs[l_str] > 1 and occs[r_str] == 1:
-                    dist = l_str
-                    target = (left, right)
-                elif occs[r_str] > 1 and occs[l_str] == 1:
-                    dist = r_str
-                    target = (right, left)
-                elif occs[l_str] == 1 and occs[r_str] == 1:
-                    dist = l_str
-                    target = (left, right)
-                else:
-                    raise RuntimeError("Distribute error: symbol not found")
-                to_distr[dist].append(target)
-
-        # Expansion ensures the expression to be in a form like:
-        # tensor[i][j] += A[i]*B[j] + C[i]*D[j] + A[i]*E[j] + ...
-        if not self._expanded:
-            raise RuntimeError("Distribute error: expansion required first.")
+            A[i]*(B[j] + C[j]).
 
-        to_distr = defaultdict(list)
-        occurrences = count_occurrences(self.stmt.children[1], key=2)
-        find_prod(self.stmt.children[1], occurrences, to_distr)
+        :param mode: multiple factorization strategies are possible, each exposing
+                     different, "hidden" opportunities for code motion.
 
-        # Create the new expression
-        new_prods = []
-        for d in to_distr.values():
-            dist, target = zip(*d)
-            target = Par(ast_c_sum(target)) if len(target) > 1 else ast_c_sum(target)
-            new_prods.append(Par(Prod(dist[0], target)))
-        self.stmt.children[1] = Par(ast_c_sum(new_prods))
+                     * mode == 'standard': this simple heuristics consists of \
```

Comment:
Are you sure? I remember we (actually you, long time ago) put them in an analogous case to fix the layout (see docstring of rewrite() in optimizer.py)